xarray/coding/variables.py

```diff
@@ -237,7 +237,7 @@ def _choose_float_dtype(dtype, has_offset):
         # Sensitivity analysis can be tricky, so we just use a float64
         # if there's any offset at all - better unoptimised than wrong!
         if not has_offset:
-            return np.float32
+            return np.float64
```

Comment:
I think the code matches the comments.  It would be clearer if written as 
```python
if has_offset:
    return np.float64
else:
    return np.float32
```

Without your edits, if there is an offset the condition does not trigger and we return np.float64 later

Reply:
Thanks for reviewing this pull request. FYI my original comment (later edited) said:

> Also, before this is merged, I'd like to suggest a larger change, and possibly discuss architecture here a bit (if appropriate). Specifically, I'd like to change the _choose_float_dtype function, and the two calls to it, to pass in the dtype of scale_factor and add_offset, in addition to the data dtype. This function should then return the dtype of the highest precision of three.
> 
> Currently, _choose_float_dtype returns float32 if the data is float16 or float32, even if the scale_factor dtype is float64.


Based on your comment, I think my original intuition - that this function needs a large rewrite - is correct. I'll look into this and submit additional commits to this PR.

Reply:
Thanks for looking into this!